### DISCUSSION TITLE
Resolving "ca md too weak" by setting SECLEVEL to 0 - what are the consequences?

### DISCUSSION THREAD

**Discussion by jgoeres** [openssl/openssl#23128] [Q&A]
Hi,
we are providing a software that uses Apache HTTPD with mod_proxy_balancer and/or mod_jk as frontend to reverse-proxy traffic to a bunch of microservice. In on premise environments, Apache HTTPD is also handling TLS, and thus we are using OpenSSL.
Customers need to provide their own certificate for the hostname on which they are running this. Many customers use their own internal  CA to create these (as the hostnames are intranet only in many cases). These CAs are usually not all that up-to-date. Currently, we have  customer whose CA produces certs with sha1 algorithm, which is reject by modern OpenSSL versions with (this is from the Apache error.log):

    [Fri Dec 22 11:07:26.661144 2023] [ssl:emerg] [pid 34676:tid 484] SSL Library Error: error:0A00018E:SSL routines::ca md too weak

The problem is clear, SHA1 is not considered secure anymore and we communicated clearly to the customer that he needs to get an SHA256 or SHA384 certificate. Alas, like in any larger organization, the process to "fix" the customer's CA to produce certificates with a secure digest algorithm are quite involved, so for the meantime, I was looking for a workaround.
I stumbled upon OpenSSL's security level setting, which can also be set through Apache's mod_ssl SSLCipherSuites directive, by appending 
    
    ":@SECLEVEL=0" 

With this setting added, Apache will start up fine with the "bad certificate"
However, I am not sure about any additional consequences of this setting, in particular if this will also impact the available ciphers that might be used for transport encryption (i.e., that the setting might not only "relax" the accepted certificate digest algorithms). The documentation I found (mainly https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_security_level.html ) does not clearly distinguish between certificate digest and transport cipher algorithms here.

So I used nmap's ssl-enum-ciphers script to check the available transport cipher protocols, which can be run like this:

    nmap --script ssl-enum-ciphers -p <PORT> <HOSTNAME>

This is the result with a "good" certificate and with our standard list of cipher suites, _without_ @SECLEVEL=0:

```
| ssl-enum-ciphers:
|   TLSv1.2:
|     ciphers:
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (ecdh_x25519) - A
|       TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 (ecdh_x25519) - A
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (ecdh_x25519) - A
|     compressors:
|       NULL
|     cipher preference: server
|   TLSv1.3:
|     ciphers:
|       TLS_AKE_WITH_AES_256_GCM_SHA384 (ecdh_x25519) - A
|       TLS_AKE_WITH_CHACHA20_POLY1305_SHA256 (ecdh_x25519) - A
|       TLS_AKE_WITH_AES_128_GCM_SHA256 (ecdh_x25519) - A
|     cipher preference: server
|_  least strength: A
```

This is the result with the "bad" (=SHA1-only) certificate and SECLEVEL=0 to actually make it start:

```
| ssl-enum-ciphers:
|   TLSv1.2:
|     ciphers:
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (ecdh_x25519) - F
|       TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 (ecdh_x25519) - F
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (ecdh_x25519) - F
|     compressors:
|       NULL
|     cipher preference: server
|     warnings:
|       Insecure certificate signature (SHA1), score capped at F
|   TLSv1.3:
|     ciphers:
|       TLS_AKE_WITH_AES_256_GCM_SHA384 (ecdh_x25519) - A
|       TLS_AKE_WITH_CHACHA20_POLY1305_SHA256 (ecdh_x25519) - A
|       TLS_AKE_WITH_AES_128_GCM_SHA256 (ecdh_x25519) - A
|     cipher preference: server
|_  least strength: F
```
It shows the same list of transport ciphers, but complains about the insecure certificate signature and therefore "downgrades" the cipher rating for TLS 1.2 from A to F. (interestingly, it does not "downgrade" the ratings of the ciphers for TLS 1.3... why?)
So this seems to indicate that we did not compromise the transport ciphers that would be used - but can this evaluation be trusted?
Even if the list of accepted transport ciphers is not "weakened", are there any other consequences related to the seclevel=0 setting?

THX in advance for any insight!

PS: Maybe someone in this community can also answer why nmap's ssl-enum-ciphers does not also downgrade the rating for the TLS 1.3 ciphers? It is the same "bad" server certificate, after all. Is this somehow a well-founded assessement or just a glitch in nmap?

**Comment by mattcaswell** [comment]
> The documentation I found (mainly https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_security_level.html ) does not clearly distinguish between certificate digest and transport cipher algorithms here.

The security level applies to all algorithms used - whether that is transport cipher algorithms or digests used in certificates or signatures in TLS messages (sigalgs).

Broadly at security level 2 the minimum number of security bits is 112 (OpenSSL 3.2), at level 1 the minimum is 80 (OpenSSL 3.0/3.1), and at 0 the minimum is 0 (legacy OpenSSL <1.0.2). There are a few other specific restrictions (not security bits related) that are specifically called out on the man page - but otherwise the minimum security bits applies.

How many security bits any particular algorithm has is a bit complicated and depends on the context to some extent. But for the purposes of X509 certificates SHA1 is assumed to have 63 bits of security with OpenSSL:

https://github.com/openssl/openssl/blob/2462e431ffe75027f253d8f1aab44ba09129c628/crypto/x509/x509_set.c#L231-L284

More generally for security level 1 this means you could use 1024 bit RSA keys, or 160 bit EC keys (actually even this is too small), but with level 0 you could use RSA keys which are far too small (e.g. 512). These days RSA keys of 2048 bits or above are generally recommended which is security level 2 (the default in OpenSSL 3.2).

A security level of 0 will also allow some ciphersuites that would not otherwise be allowed such as integrity only ciphersuites (i.e. NULL encryption), those based around CCM8 and the so-called "weak" ciphers which are only available in OpenSSL if explicitly compiled in.

However, what is actually *used* in a connection is a product of what the *configuration* allows and what the *security level* allows. So if the configuration doesn't allow any weak ciphersuites then having a lower security level won't have an impact. The security level is there to prevent "foot-gun" configurations. If you lower the security level you are opening up the potential for mistakes - so you better be really sure that your configuration is good.

**Reply by jgoeres** [reply]
First: Thanks for the fast and detailed response!

You mention these "areas of impact" of the seclevel:

- "transport cipher algorithms" - from what I gathered, assuming my list of accepted cipher suites I have in my Apache SSLCipherSuites directives is fine, it appears acceptable to set the SECLEVEL to 0
- "digests used in certificates" - here it is also fine, cause this is what we want
- "signatures in TLS messages" - I am not fully sure on what could be security/confidentiality implications here. From what little I remember of the details of SSL/TLS, could this open gaps in the initial TLS handshake?

THX

J

**Reply by mattcaswell** [reply]
> You mention these "areas of impact" of the seclevel:

Note that I was providing some examples - not an exhaustive list. Probably I should add to that list key exchange algs/supported groups. E.g. see

https://www.openssl.org/docs/man3.2/man3/SSL_CTX_set1_groups_list.html

> "signatures in TLS messages" - I am not fully sure on what could be security/confidentiality implications here. From what little I remember of the details of SSL/TLS, could this open gaps in the initial TLS handshake?

Right. For example it would be possible to have a certificate signed using the SHA256 digest, but the signatures generated from its public key in the TLS handshake could end up only using SHA1. E.g. see:

https://www.openssl.org/docs/man3.2/man3/SSL_CTX_set1_sigalgs_list.html

**Reply by jgoeres** [reply]
Thanks again for the detailed info about the security level. Due to this setting's broad impact, for me, the summary is that it is probably not a good idea to lower the security level to 0 "just" to allow a certificate with an insecure hash algorithm. 
Which brings up the next question: is there an option to allow only the use of such an "insecure" certificate, but without opening up all kinds of other security issues that sec level 0 brings with it?

**Reply by mattcaswell** [reply]
It is possible to do by making application changes (e.g. I assume mod_ssl for you) - but it is not possible via configuration. That might be a nice feature to have.

To do it by application changes, the application would need to supply a custom security callback. It is the default security callback that currently implements these restrictions. By implementing a custom security callback you could intercept the checks you are interested in, and proxy all other checks to the default security callback.

See in particular the documentation for the SSL_CTX_get_security_callback() and SSL_CTX_set_security_callback() functions (and their `SSL` equivalents):

https://www.openssl.org/docs/man3.2/man3/SSL_CTX_set_security_callback.html

The man page, under "application defined callbacks" has the "helpful" note: "Documentation to be provided"!!

**Reply by kidharb** [reply]
@mattcaswell I would like to complete the "application defined callbacks" documentation, I will raise a pull request

**Reply by mattcaswell** [reply]
> @mattcaswell I would like to complete the "application defined callbacks" documentation, I will raise a pull request

Awesome! Thanks.
